taichi/transforms/make_adjoint.cpp

```diff
@@ -1,62 +1,33 @@
 #include <typeinfo>
 #include "taichi/ir/ir.h"
-#include <taichi/ir/frontend.h>
+#include "taichi/ir/frontend.h"
 
 TLANG_NAMESPACE_BEGIN
 
-template <typename T>
-class StmtSearcher : public BasicStmtVisitor {
- private:
-  std::function<bool(Stmt *)> test;
-  std::vector<Stmt *> results;
-
- public:
-  using BasicStmtVisitor::visit;
-
-  StmtSearcher(std::function<bool(Stmt *)> test) : test(test) {
-    allow_undefined_visitor = true;
-    invoke_default_visitor = true;
-  }
-
-  void visit(Stmt *stmt) {
-    if (stmt->is<T>() && test(stmt))
-      results.push_back(stmt);
-  }
-
-  static std::vector<Stmt *> run(IRNode *root,
-                                 std::function<bool(Stmt *)> test) {
-    StmtSearcher<T> searcher(test);
-    root->accept(&searcher);
-    return searcher.results;
-  }
-};
-
 // Do automatic differentiation pass in the reverse order (reverse-mode AD)
 
 class ConvertLocalVar : public BasicStmtVisitor {
  public:
   using BasicStmtVisitor::visit;
 
-  void visit(AllocaStmt *stmt) override {
-    TI_ASSERT(stmt->width() == 1);
-    bool no_store = StmtSearcher<LocalStoreStmt>::run(
-                        stmt->parent,
-                        [&](Stmt *store) {
-                          return store->cast<LocalStoreStmt>()->ptr == stmt;
-                        })
-                        .empty();
-    bool no_atomic = StmtSearcher<AtomicOpStmt>::run(
-                         stmt->parent,
-                         [&](Stmt *atomic) {
-                           return atomic->cast<AtomicOpStmt>()->dest == stmt;
+  void visit(AllocaStmt *alloc) override {
+    TI_ASSERT(alloc->width() == 1);
+    bool load_only = irpass::gather_statements(
+                         alloc->parent,
```

Comment:
`alloc->parent` means we gather statements not only before `alloc` but also after `alloc`. Since it's an `AllocaStmt`, it looks equivalent to only gather statements after `alloc`. I believe the efficiency here doesn't affect much though.

Reply:
Right, I mean you can simplify your `LocalLoadSearcher` into a few lines of code with `gather_statements(...).empty()`

Reply:
To only test if there is at least one specific statement (which can break/return whenever we find one), or to gather all statements, will they make a difference in efficiency?

Reply:
You are right that we can probably optimize this to enable a short-cut (return once find) in the future. Here just a generic implementation for simplicity.